jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/viewmodel/AudioDataStore.kt

```diff
@@ -0,0 +1,176 @@
+            audioPlayer.loadSection(source.file, source.start, source.end)
+            sourceAudioPlayerProperty.set(audioPlayer)
+        } ?: sourceAudioPlayerProperty.set(null)
+    }
+
+    fun closeSourceAudioPlayer() {
+        sourceAudioPlayerProperty.value?.close()
+    }
+
+    fun stopSourceAudioPlayer() {
+        sourceAudioPlayerProperty.value?.stop()
+    }
+
+    fun openTargetAudioPlayer() {
+        targetAudioProperty.value?.let { target ->
+            target.player.load(target.file)
+        }
+    }
+
+    fun closePlayers() {
+        selectedChapterPlayerProperty.value?.close()
+        targetAudioProperty.value?.player?.close()
+        sourceAudioPlayerProperty.value?.close()
+        sourceAudioPlayerProperty.set(null)
+    }
+
+    fun stopPlayers() {
+        sourceAudioPlayerProperty.value?.stop()
+        targetAudioProperty.value?.player?.stop()
+    }
+
+    fun updateSelectedChapterPlayer() {
+        val chunk = workbookDataStore.activeChunkProperty.value
+        val chapter = workbookDataStore.activeChapterProperty.value
+        when {
+            chapter != null && chunk == null -> {
+                val take = chapter.audio.selected.value?.value
+                take?.let {
+                    updateTargetAudio(it.file)
+
+                    val audioPlayer = getNewAudioPlayer()
+                    audioPlayer.load(it.file)
+                    selectedChapterPlayerProperty.set(audioPlayer)
+                } ?: run {
+                    selectedChapterPlayerProperty.set(null)
+                    targetAudioProperty.set(null)
+                }
+            }
+            chapter != null -> { /* no-op */
+            } // preserve targetAudio for clean up
+            else -> {
+                selectedChapterPlayerProperty.set(null)
+                targetAudioProperty.set(null)
+            }
+        }
+    }
+
+    private fun updateTargetAudio(file: File) {
+        cleanUpTargetAudio()
+
+        val tempFile = directoryProvider.createTempFile(
+            file.nameWithoutExtension,
+            ".${file.extension}"
+        )
+        file.copyTo(tempFile, true)
+
+        val audioPlayer = getNewAudioPlayer()
+        audioPlayer.load(tempFile)
+        val targetAudio = TargetAudio(tempFile, audioPlayer)
+
+        targetAudioProperty.set(targetAudio)
+    }
+
+    private fun cleanUpTargetAudio() {
+        targetAudioProperty.value?.let {
+            it.player.release()
+            it.file.delete()
+        }
+        targetAudioProperty.set(null)
+    }
+
+    fun updateSourceAudio() {
+        workbookDataStore.activeWorkbookProperty.value?.let { workbook ->
+            val chunk = workbookDataStore.activeChunkProperty.get()
+            val chapter = workbookDataStore.activeChapterProperty.get()
+            if (chapter != null && chunk != null) {
+                sourceAudioProperty.set(workbook.sourceAudioAccessor.getChunk(chapter.sort, chunk.sort, workbook.target))
```

Comment:
when a method call is getting long and nested like this, I usually introduce a variable so that it's easier to know what value is set

```
val ___ = workbook.sourceAudioAccessor.getChunk(chapter.sort, chunk.sort, workbook.target)
sourceAudioProperty.set(___)
```